File: app/javascript/controllers/checkout_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["monthlyCheckout", "annualCheckout"]
  connect() {
    Paddle.Environment.set('sandbox');
    Paddle.Setup({
      seller: 9943,
      eventCallback: function(data) {
        if (data.name == "checkout.completed") {
          console.log(data)
          let createdBusinessId

          // Saving the below variables to session data to use in my account area
          let createdCustomerId = data.data.customer.id;
          console.log(createdCustomerId)
          let createdAddressId = data.data.customer.address.id;
          console.log(createdAddressId)
          let createdCheckoutId = data.data.id;
          console.log(createdCheckoutId)
          let createdTransactionId = data.data.transaction_id;
          console.log(createdTransactionId)
          if (data.data.customer.business) {
            createdBusinessId = data.data.customer.business.id
            console.log(createdBusinessId)
          } else {
            console.log("No business info provided")
          }
          let createdQuantity = data.data.items[0].quantity

          let params = new URLSearchParams();
          params.append("createdCustomerId", createdCustomerId)
          params.append("createdAddressId", createdAddressId)
          params.append("createdCheckoutId", createdCheckoutId)
          params.append("createdTransactionId", createdTransactionId)
          if (typeof createdBusinessId !== 'undefined') {
            params.append("createdBusinessId", createdBusinessId)
          } else {
            console.log("No business info provided")
          }
          params.append("createdQuantity", createdQuantity)

          fetch("/save_customer?" + params)
            .then(response => response.json())
            .then(data => {
              console.log(data)
            })
            .catch(error => {
              console.error("Error saving customer", error);
            });
        }
      }
    });
  }

  openCheckout() {
    let monthlyQuantity = document.getElementById("monthly-users").value;
    let monthlyCheckboxValue = document.getElementById("monthly-onboarding-support").checked;

    let params = new URLSearchParams();
    params.append("monthlyQuantity", monthlyQuantity);
    params.append("monthlyCheckboxValue", monthlyCheckboxValue);

    fetch("/create_monthly_transaction?" + params)
      .then(response => response.json())
      .then(data => {
        let createdTransactionId = data.data.id
        console.log(createdTransactionId)

        Paddle.Checkout.open({
          settings: {
            theme: "light"
          },
          transactionId: createdTransactionId,
          // discountId: "dsc_01h2bnv663v06e2g65cx8crg8m",
        });
      })
      .catch(error => {
        console.error("Error creating transaction:", error);
      });
  }
}

File: app/controllers/checkout_controller.rb
require 'uri'
require 'net/http'
require 'openssl'

class CheckoutController < ApplicationController
  def save_customer
    @customer_id = params[:createdCustomerId]
    @address_id = params[:createdAddressId]
    @checkout_id = params[:createdCheckoutId]
    @transaction_id = params[:createdTransactionId]
    @quantity = params[:createdQuantity]

    session[:session_customer_id] = @customer_id
    session[:session_address_id] = @address_id
    session[:session_checkout_id] = @checkout_id
    session[:session_transaction_id] = @transaction_id

    if params[:createdBusinessId].present?
      @business_id = params[:createdBusinessId]
      session[:session_business_id] = @business_id
    end

    session[:session_quantity] = @quantity

    response = {
      status: "success",
      message: "Created IDs saved to session storage!"
    }

    render json: response, content_type: 'application/json'
  end

  # This method below was going to be used to get all info on the customer, subscription etc and could be used by other controllers
  # def get_all_info
  #   {
  #     "customer_id" => @customer_id,
  #     "address_id" => @address_id,
  #     "checkout_id" => @checkout_id,
  #     "transaction_id" => @transaction_id,
  #     "business_id" => @business_id
  #   }
  # end
end
